### PR TITLE
Overwrite mode

### PR DESCRIPTION
**Pull request by slewis74**
Adding OverwriteMode, to replace the `replace` boolean in the push APIs. Adds the ability for the caller to IgnoreExisting.

Relates to OctopusDeploy/Octopus-TeamCity#18

### CODE REVIEW

File: source/Octopus.Client/Repositories/BuiltInPackageRepositoryRepository.cs [hunk 1]
```diff
@@ -57,18 +69,32 @@ public PackageFromBuiltInFeedResource PushPackage(string fileName, Stream conten
             }
                 
             contents.Seek(0, SeekOrigin.Begin);
+
+            var link = repository.Link("PackageUpload");
+            object pathParameters;
+
+            // if the link doesn't contain overwritemode then we're connected to an older server, which uses the `replace` parameter  
+            if (link.Contains("overwritemode"))
```
Comment (hnrkndrssn) [inline source/Octopus.Client/Repositories/BuiltInPackageRepositoryRepository.cs]: Should this be using `OverwriteModeLink.Link`?

File: source/Octopus.Client/Repositories/Async/BuiltInPackageRepositoryRepository.cs [hunk 2]
```diff
@@ -62,18 +71,31 @@ public async Task<PackageFromBuiltInFeedResource> PushPackage(string fileName, S
                 Logger.Info("Pushing the complete package to the server, as delta compression was explicitly disabled");
             }
 
+            var link = await repository.Link("PackageUpload").ConfigureAwait(false);
+            object pathParameters;
+
+            // if the link doesn't contain overwritemode then we're connected to an older server, which uses the `replace` parameter  
+            if (link.Contains("overwritemode"))
+            {
+                pathParameters = new {replace = overwriteMode == OverwriteMode.OverwriteExisting};
```
Comment (hnrkndrssn) [inline source/Octopus.Client/Repositories/Async/BuiltInPackageRepositoryRepository.cs]: This logic looks opposite to what I'd expect, i.e. if server support `overwriteMode` we should pass in `overwriteMode` in the path params...

File: source/Octopus.Client/Repositories/Async/BuiltInPackageRepositoryRepository.cs [hunk 3]
```diff
@@ -62,18 +71,31 @@ public async Task<PackageFromBuiltInFeedResource> PushPackage(string fileName, S
                 Logger.Info("Pushing the complete package to the server, as delta compression was explicitly disabled");
             }
 
+            var link = await repository.Link("PackageUpload").ConfigureAwait(false);
+            object pathParameters;
+
+            // if the link doesn't contain overwritemode then we're connected to an older server, which uses the `replace` parameter
```
Comment (hnrkndrssn) [inline source/Octopus.Client/Repositories/Async/BuiltInPackageRepositoryRepository.cs]: This comment makes me think you initially wanted to do `if (!link.Contains)` here...?

File: source/Octopus.Client/Repositories/BuiltInPackageRepositoryRepository.cs [hunk 4]
```diff
@@ -57,18 +69,32 @@ public PackageFromBuiltInFeedResource PushPackage(string fileName, Stream conten
...
+            {
+                pathParameters = new {replace = overwriteMode == OverwriteMode.OverwriteExisting};
```
Comment (hnrkndrssn) [inline source/Octopus.Client/Repositories/BuiltInPackageRepositoryRepository.cs]: Looks like the logic is reversed here too

File: source/Octopus.Client/Repositories/Async/BuiltInPackageRepositoryRepository.cs [hunk 5]
```diff
@@ -62,18 +71,31 @@ public async Task<PackageFromBuiltInFeedResource> PushPackage(string fileName, S
                 Logger.Info("Pushing the complete package to the server, as delta compression was explicitly disabled");
             }
 
+            var link = await repository.Link("PackageUpload").ConfigureAwait(false);
+            object pathParameters;
+
+            // if the link doesn't contain overwritemode then we're connected to an older server, which uses the `replace` parameter  
+            if (link.Contains("overwritemode"))
```
[the remark this answers is not in the document]
Comment (thedewi) [reply source/Octopus.Client/Repositories/Async/BuiltInPackageRepositoryRepository.cs]: Agree, it should be.

File: source/Octopus.Client/Repositories/Async/BuiltInPackageRepositoryRepository.cs (hunk 2 above)
Comment (thedewi) [reply source/Octopus.Client/Repositories/Async/BuiltInPackageRepositoryRepository.cs]: Agree, most of these blocks have the `overwriteMode` case first, and this one should too.

File: source/Octopus.Client/Repositories/BuiltInPackageRepositoryRepository.cs (hunk 1 above)
Comment (thedewi) [reply source/Octopus.Client/Repositories/BuiltInPackageRepositoryRepository.cs]: Agree.

File: source/Octopus.Client/Repositories/BuiltInPackageRepositoryRepository.cs (hunk 4 above)
Comment (thedewi) [reply source/Octopus.Client/Repositories/BuiltInPackageRepositoryRepository.cs]: Agree, most of these have the `overwriteMode` case first, so I recommend swapping this one.

**Comment by thedewi** [comment]
fixed those 4 issues in #439

**Pull request merged**